xarray/core/dataarray.py

```diff
@@ -421,22 +422,29 @@ def __init__(
         # public interface.
         self._indexes = indexes
 
-        self._file_obj = None
+        self._close = None
 
     def _replace(
         self,
         variable: Variable = None,
         coords=None,
         name: Union[Hashable, None, Default] = _default,
         indexes=None,
+        close: Union[Callable[[], None], None, Default] = _default,
     ) -> "DataArray":
         if variable is None:
             variable = self.variable
         if coords is None:
             coords = self._coords
         if name is _default:
             name = self.name
-        return type(self)(variable, coords, name=name, fastpath=True, indexes=indexes)
+        if close is _default:
+            close = self._close
+        replaced = type(self)(
+            variable, coords, name=name, fastpath=True, indexes=indexes
+        )
+        replaced.set_close(close)
+        return replaced
```

Comment:
I would prefer to revert this change to `_replace`, and require users to always explictly call `set_close` rather than to update it automatically in only some cases.

Reply:
Not sure I understand: `_close` follows the same rules as any other `Dataset` / `DataArray` attribute, it is updated if passed otherwise it is copied from `self`.

Note that this change is the main reason the PR is so small. If I remove this I'll need to hunt down all the calls to `_replace` and choose whether to add or not to add `set_close`.

I may not have understood your concern.

Reply:
> Not sure I understand: _close follows the same rules as any other Dataset / DataArray attribute, it is updated if passed, otherwise it is copied from self.

I'm suggesting that `_close` should be an exception to this rule, and that we should never call `set_close()` inside `_replace()`. Anytime `_replace()` is called (creating a new `DataArray`), the right thing to do is to lose the `close` method.

> If I remove this I'll need to hunt down all the calls to _replace and choose whether to add or not to add set_close.

My suggestion is that it is OK to *never* add `set_close` (unless inside a function like `open_dataset` or `open_mfdataset` that opens file object(s) and already assigns to `_file_obj`)

Reply:
So the `Dataset.close` method and `DataArray.close` method are supposed to actually release the resources only when called from the objects returned directly from the `xr.open_*` functions?

```python
>>> ds1 = xr.open_dataset("file.nc")
>>> ds1.close()  # releases the resources

>>> ds2 = xr.open_dataset("file.nc").rename_dims(valid_time="time")
>>> ds2.close()  # doesn't release the resources
```

I don't understand the rationale, but I'll trust your judgment on this.

Reply:
Done, this is ready for me.

Reply:
@shoyer This is really great and helps a lot in resource handling. Your example should be put into documentation.